Dynamic build support with CI matrix

angerman opened the pull request:
## Summary

This PR adds comprehensive dynamic build support to the stable-haskell GHC build, building on the foundation from PR #137.

### Changes

**CI/Build System:**
- Add DYNAMIC=0/1 build/test matrix for static and dynamic builds
- Add basename symlinks for nested shared libraries (.dylib, .so) to make them discoverable by the linker
- Create ghc-iserv-dyn symlink so GHC can find the dynamic interpreter
- Use concrete file target for testsuite-timeout to avoid unnecessary rebuilds

**RTS/Linker:**
- Improve RTS sublibrary support with -no-ghc-internal flag
- Fix dynamic linking for RTS sublibraries (rpath injection, RTLD_GLOBAL for ghc-internal)
- Dynamically compute RTS rpath dependencies instead of hardcoded list
- Export RTS symbols from ghc-iserv for dynamic builds
- Fix libffi-clib integration for dynamic builds

**Testsuite:**
- Improve testlib.py robustness with better error handling
- Adjust tests for RTS split

### Based on
- PR #137 by @hasufell (Build dynamic GHC in github CI)

## Test plan
- [x] CI passes for both DYNAMIC=0 and DYNAMIC=1 builds
- [x] Testsuite passes on all platforms

angerman commented:
Interestingly, this now seems to fail due to the missing rts sublib issues. That are fixed in #44.

angerman commented on testsuite/mk/boilerplate.mk:
I should add a comment here why we do this. It adds a hash to the ghc config file. This ensure we always read the correct ghcconfig, if the ghc changed, the hash changed, and we recompute.

angerman commented on testsuite/driver/testlib.py:
the `dsos` should not be in here anymore.

angerman commented on rts/rts.cabal:
Julian has "rts_v" and others in #75 for this.

hsyl20 commented on compiler/GHC/Driver/Session.hs:
Why is this needed now?

angerman commented on .gitignore:
This should be in a separate PR.

angerman commented on cabal.project.stage1:
We should document why this is needed. I think there were some warnings from the libffi-clib, which then became errors and just failed the build.

angerman commented on Makefile:
Why did we change this? This seems like we don't want this change. The existing logic seems fine.

angerman commented on compiler/GHC/Driver/Flags.hs:
Do we still need the no-ghc-internal flag? 🤔

angerman commented on Makefile:
This is much better tracking instead of some phony stuff.

angerman commented on compiler/GHC/Driver/Flags.hs:
Yes we do.

angerman commented on .gitignore:
It's in #120 now.

hasufell commented on rts/configure.ac:
`mkdir -p` is not terribly portable. We can use the `AC_PROG_MKDIR_P` macro and then use `${MKDIR_P} include/rts`.

Also see https://www.gnu.org/software/autoconf/manual/autoconf-2.68/html_node/Particular-Programs.html

hasufell commented on Makefile:
What is this? Why is this needed?

hasufell commented on Makefile:
for loops with find are dirty... prefer `-exec` or `-execdir`

angerman commented on Makefile:
Good idea!

angerman commented on Makefile:
This is something we should ultimately fix in cabal; we do not want shared objects to be in various subdirectories,as that requires ballooning the executable with RPATHs for each subdirecty. We want dynamic lirbaries to live in _one_ directory.

angerman commented on rts/configure.ac:
Yes, good idea! Thank you!

angerman commented:
There is now #137

hasufell commented:
On wasm I get with these changes:

```
RtsStartup.c:106:5: error:
     error: use of undeclared identifier 'Dl_info'
      106 |     Dl_info info;
          |     ^~~~~~~
    |
106 |     Dl_info info;
    |     ^

RtsStartup.c:107:9: error:
     error: call to undeclared function 'dladdr'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
      107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
          |         ^
    |
107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
    |         ^

RtsStartup.c:107:44: error:
     error: use of undeclared identifier 'info'
      107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
          |                                            ^~~~
    |
107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
    |                                            ^

RtsStartup.c:107:53: error:
     error: use of undeclared identifier 'info'
      107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
          |                                                     ^~~~
    |
107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
    |                                                     ^

RtsStartup.c:108:31: error:
     error: use of undeclared identifier 'info'
      108 |         void* handle = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL);
          |                               ^~~~
    |
108 |         void* handle = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL);
    |                               ^
```

angerman commented:
@hasufell yes, on WASM `dladdr` doesn't exist. Let's defer WASM for now, until we get to the root of the stupid linking issues. Latest I saw was ghc-iserv loading libc twice. Something is off... somewhere.

> On wasm I get with these changes:
> 
> ```
> RtsStartup.c:106:5: error:
>      error: use of undeclared identifier 'Dl_info'
>       106 |     Dl_info info;
>           |     ^~~~~~~
>     |
> 106 |     Dl_info info;
>     |     ^
> 
> RtsStartup.c:107:9: error:
>      error: call to undeclared function 'dladdr'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>       107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
>           |         ^
>     |
> 107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
>     |         ^
> 
> RtsStartup.c:107:44: error:
>      error: use of undeclared identifier 'info'
>       107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
>           |                                            ^~~~
>     |
> 107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
>     |                                            ^
> 
> RtsStartup.c:107:53: error:
>      error: use of undeclared identifier 'info'
>       107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
>           |                                                     ^~~~
>     |
> 107 |     if (dladdr((void*)&init_ghc_hs_iface, &info) && info.dli_fname) {
>     |                                                     ^
> 
> RtsStartup.c:108:31: error:
>      error: use of undeclared identifier 'info'
>       108 |         void* handle = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL);
>           |                               ^~~~
>     |
> 108 |         void* handle = dlopen(info.dli_fname, RTLD_NOW | RTLD_NOLOAD | RTLD_GLOBAL);
>     |                               ^
> ```

hasufell commented:
On ARM64 alpine musl linked:

```
2026-02-05T06:41:54.1525196Z +++ ghci/linking/dyn/T10458.run/T10458.run.stderr.normalised	2026-02-05 05:56:23.588129109 +0000
2026-02-05T06:41:54.1525572Z @@ -0,0 +1,11 @@
2026-02-05T06:41:54.1525710Z +
2026-02-05T06:41:54.1525937Z +GHC.Linker.Loader.dynLoadObjs: Loading temp shared object failed
2026-02-05T06:41:54.1526334Z +During interactive linking, GHCi couldn't find the following symbol:
2026-02-05T06:41:54.1526878Z +  Error loading shared library libAS.so: No such file or directory (needed by /tmp/ghc65945_tmp_9_0/libghc_tmp_10.so)
2026-02-05T06:41:54.1527408Z +This may be due to you not asking GHCi to load extra object files,
2026-02-05T06:41:54.1527802Z +archives or DLLs needed by your current session.  Restart GHCi, specifying
2026-02-05T06:41:54.1528228Z +the missing library using the -L/path/to/object/dir and -lmissinglibname
2026-02-05T06:41:54.1528641Z +flags, or simply by naming the relevant files on the GHCi command line.
2026-02-05T06:41:54.1529026Z +Alternatively, this link failure might indicate a bug in GHCi.
2026-02-05T06:41:54.1529380Z +If you suspect the latter, please report this as a GHC bug:
2026-02-05T06:41:54.1529698Z +  https://github.com/stable-haskell/ghc/issues
```

angerman commented:
> On ARM64 alpine musl linked:
> 
> ```
> 2026-02-05T06:41:54.1525196Z +++ ghci/linking/dyn/T10458.run/T10458.run.stderr.normalised	2026-02-05 05:56:23.588129109 +0000
> 2026-02-05T06:41:54.1525572Z @@ -0,0 +1,11 @@
> 2026-02-05T06:41:54.1525710Z +
> 2026-02-05T06:41:54.1525937Z +GHC.Linker.Loader.dynLoadObjs: Loading temp shared object failed
> 2026-02-05T06:41:54.1526334Z +During interactive linking, GHCi couldn't find the following symbol:
> 2026-02-05T06:41:54.1526878Z +  Error loading shared library libAS.so: No such file or directory (needed by /tmp/ghc65945_tmp_9_0/libghc_tmp_10.so)
> 2026-02-05T06:41:54.1527408Z +This may be due to you not asking GHCi to load extra object files,
> 2026-02-05T06:41:54.1527802Z +archives or DLLs needed by your current session.  Restart GHCi, specifying
> 2026-02-05T06:41:54.1528228Z +the missing library using the -L/path/to/object/dir and -lmissinglibname
> 2026-02-05T06:41:54.1528641Z +flags, or simply by naming the relevant files on the GHCi command line.
> 2026-02-05T06:41:54.1529026Z +Alternatively, this link failure might indicate a bug in GHCi.
> 2026-02-05T06:41:54.1529380Z +If you suspect the latter, please report this as a GHC bug:
> 2026-02-05T06:41:54.1529698Z +  https://github.com/stable-haskell/ghc/issues
> ```

I'll look at this, why are we running dyn on static musl/alpine? That's odd, but probably something we should either prohibit by construction or fix.

angerman committed:
rts: fix RTS linker for dynamic GHC builds

This commit fixes several critical issues with the RTS object linker that
prevented dynamic GHC builds from loading code correctly.

Key fixes:

1. Detect data vs code references in X86_64_ELF_NONPIC_HACK (Elf.c)
   - The jump island mechanism was incorrectly applied to data references
   - For info table pointers (_con_info symbols), embedding a jump island
     address caused GC crashes ("strange closure type")
   - Now distinguishes R_X86_64_PLT32 (code) from R_X86_64_PC32 (data)
   - Data references use GOT-style indirection through extra->addr instead

2. Preserve dlerror for linker script fallback handling (LoadNativeObjPosix.c)
   - dlerror() clears after first call, losing error context
   - Now saves error string before retry logic
   - Fixes misleading error messages when loading fails

3. Promote boot libraries to RTLD_GLOBAL for dynamic code loading (RtsStartup.c)
   - Boot libraries loaded with RTLD_LOCAL weren't visible to dlsym
   - Dynamic object loading failed to resolve symbols from boot libs
   - Now re-opens boot libraries with RTLD_GLOBAL flag at startup

4. Skip loading libc/libm already linked into process (LoadNativeObjPosix.c)
   - Avoids redundant loading and symbol conflicts
   - Checks if library is already resident before calling dlopen

5. Dynamic lookup of stg_interp_constr entry points via dlsym (RtsSymbols.c)
   - Interpreter constructor symbols need runtime resolution
   - Adds dynamic fallback when static symbols unavailable

6. Remove residual debug instrumentation
   - Cleans up debugging code from Evac.c and LoadNativeObjPosix.c

angerman committed:
rts: add dynamic linking support for RTS sublibraries

This commit adds infrastructure for RTS sublibrary loading in dynamic builds,
enabling the split RTS architecture to work with shared library linking.

Key changes:

1. RTS sublibrary infrastructure (rts/rts.cabal)
   - Define separate sublibraries for RTS components
   - Add proper library dependencies and visibility
   - Configure shared library generation for RTS parts

2. Configure support for dynamic builds (rts/configure.ac)
   - Detect platform-specific dynamic linking requirements
   - Set appropriate linker flags for each sublibrary
   - Handle symbol visibility for exported functions

3. API updates for sublibrary boundaries (rts/include/RtsAPI.h)
   - Adjust exported symbol declarations
   - Ensure proper visibility across sublibrary boundaries

4. AutoApply support for interpreter (rts/AutoApply*.cmm)
   - Add AutoApply.cmm and vector variants (V16, V32, V64)
   - Required for dynamic bytecode interpreter operation

5. Cabal project configuration
   - cabal.project.stage1: Add no-ghc-internal flag for stage1 builds
   - cabal.project.stage2: Configure full RTS with all sublibraries

6. Thread infrastructure (rts/Threads.h)
   - Updates for sublibrary thread handling

angerman committed:
compiler: fixes for dynamic GHC builds

This commit addresses compiler and driver issues specific to dynamic GHC
builds, ensuring proper code generation and linking behavior.

Key changes:

1. Make Opt_ExternalDynamicRefs default on all PIC platforms (DynFlags.hs)
   - Previously only enabled for specific configurations
   - Dynamic builds require external dynamic references for proper GOT usage
   - Prevents relocation issues with large code models

2. Pipeline and session handling updates
   - Driver/Pipeline.hs: Handle dynamic linking in compilation pipeline
   - Driver/Session.hs: Session configuration for dynamic builds
   - Driver/Flags.hs: Flag handling for dynamic mode

3. Linker updates for dynamic mode
   - Linker/Executable.hs: Executable linking for dynamic builds
   - Linker/Static.hs: Static linking coordination
   - ByteCode/Linker.hs: Bytecode linker for dynamic interpreter

4. Unit state and GHCi support
   - Unit/State.hs: Package database handling for dynamic libs
   - GHCi/InfoTable.hsc: Info table generation for dynamic mode
   - Tc/Gen/Splice.hs: Template Haskell splice handling

angerman committed:
ghc-bin, ghc-iserv: export RTS symbols for dynamic code loading

GHC and ghc-iserv load Haskell shared libraries dynamically for Template
Haskell and GHCi. These libraries reference RTS symbols (e.g.,
stg_INTLIKE_closure) that are linked into the executable. Without special
linker flags, those symbols aren't visible to dlopen'd libraries.

This commit adds platform-specific linker flags to export these symbols:

- Linux/FreeBSD: -rdynamic (passes --export-dynamic to ld)
- macOS: -flat_namespace (makes all symbols visible across namespaces)
- Windows: Cannot use --export-all-symbols due to 65535 symbol limit

See Note [ghc-iserv and dynamic symbol export] in ghc-iserv.cabal.in
for detailed explanation of the approach and alternatives considered.

angerman committed:
build: infrastructure for DYNAMIC=1 builds

This commit adds build system support for creating dynamic GHC builds,
including Makefile targets, bindist generation, and utility configurations.

Key changes:

1. Makefile enhancements
   - Add DYNAMIC=1 build variable support
   - Create dylib symlinks for macOS dynamic builds
   - Use concrete file target for testsuite-timeout
   - Include ghc-iserv-dyn in tarballs for all targets
   - Proper bindist generation for dynamic builds

2. Utility cabal files (hp2ps.cabal, unlit.cabal)
   - Configure for dynamic linking support
   - Ensure utilities work with dynamic GHC

3. ghc-iserv infrastructure (iservmain.c)
   - Updates for dynamic interpreter server
   - Proper initialization for dynamic linking context

4. Test expectations for Stable Haskell
   - Update bug report URL in test expectations

Usage:
  make DYNAMIC=1 _build/bindist  # Build dynamic GHC bindist

angerman committed:
testsuite: adjustments for RTS split and dynamic GHC builds

This commit updates the testsuite to handle the split RTS architecture
and dynamic GHC build configuration.

Key changes:

1. testlib.py improvements
   - More robust test driver for dynamic builds
   - Better handling of shared library paths
   - Improved error detection and reporting

2. Test infrastructure (boilerplate.mk)
   - Configure tests for dynamic linking environment
   - Set proper library paths for test execution

3. Test adjustments for RTS split
   - T18072debug: Update grep to match cabal-based RTS naming
   - T23142.hs: Revert module name to fix -Di debug output test
   - keep-cafs-fail.stdout: Update expected output

4. Dynamic linking test updates
   - ghci/linking/dyn/all.T: Adjust for dynamic GHC
   - T2228: Restore expect_broken(7298) for dynamic builds
   - T11531.stderr: Update expected error messages

5. Platform-specific adjustments
   - T10458: Skip on musl with dynamic GHC
   - T11223 tests: Update stderr expectations for Windows

6. Test configuration
   - .gitignore: Add patterns for dynamic test artifacts
   - dynlibs/Makefile: Update for dynamic build testing
   - perf/size/all.T: Adjust size expectations

angerman committed:
ci: add dynamic build matrix to GitHub workflows

This commit extends the CI/CD pipeline to build and test dynamic GHC
configurations alongside the existing static builds.

Key changes:

1. ci.yml - Main CI workflow
   - Add DYNAMIC=1 to build matrix
   - Configure dynamic build jobs for Linux and macOS
   - Run ghci-ext tests on dynamic builds (require interpreter)
   - Parallel execution of static and dynamic builds

2. reusable-release.yml - Release workflow
   - Add dynamic GHC builds to release artifacts
   - Generate separate bindists for dynamic configuration
   - Include ghc-iserv-dyn in release tarballs
   - Re-enable release workflow on pull requests for testing

The dynamic build matrix allows testing of:
- Template Haskell with dynamic code loading
- GHCi interactive features
- Dynamic library loading and linking
- Interpreter-based test suites (ghci-ext)

Build configurations:
- Static (default): DYNAMIC=0 or unset
- Dynamic: DYNAMIC=1

angerman commented on Makefile:
I wonder if we really need these still?

angerman commented on Makefile:
Yuck. this is so stupid.

angerman commented on Makefile:
The need for all these LD_LIBRARY_PATH stuff, feels more like a hack. Something we ought to fix properly.

angerman commented on Makefile:
I believe this is fixed in cabal now, the copy-headers stuff?

angerman commented on Makefile:
Again, this feels like a massive hack, we shouldn't need and that GHC should just do on it's own properly.

angerman commented on utils/ghc-iserv/cbits/iservmain.c:
I don't think this comment is necessary anymore 🤔

hsyl20 commented on libraries/ghci/GHCi/InfoTable.hsc:
You don't need an IORef: a CAF would be enough. On first entry it would perform the `unsafePerformIO` stuff.

hsyl20 commented on libraries/ghci/GHCi/InfoTable.hsc:
In `rts/includes/stg/MiscClosures.h` there is this comment:

```c
// RTS_FUN(stg_interp_constr1_entry);
// RTS_FUN(stg_interp_constr2_entry);
// RTS_FUN(stg_interp_constr3_entry);
// RTS_FUN(stg_interp_constr4_entry);
// RTS_FUN(stg_interp_constr5_entry);
// RTS_FUN(stg_interp_constr6_entry);
// RTS_FUN(stg_interp_constr7_entry);
//
// This is referenced using the FFI in the compiler (GHC.ByteCode.InfoTable),
// so we can't give it the correct type here because the prototypes
// would clash (FFI references are always declared with type StgWord[]
// in the generated C code). 
```

You should now be allowed to declare the correct types here and to remove the comment I think.

hsyl20 commented on rts/linker/Elf.c:
```suggestion
 * fit in a 32-bit slot. When a reference to a symbol in a shared library
 * exceeds this range, we can use a hack (enabled by X86_64_ELF_NONPIC_HACK)
 * which consists in generating "jump islands" (trampolines) for code references
 * (this happens in makeSymbolExtra()).
*
 * For FUNCTION symbols (STT_FUNC), jump islands work correctly: the call
 * bounces through the trampoline to reach the real function.
*
 * For DATA symbols (STT_OBJECT, STT_NOTYPE), jump islands are INCORRECT:
 * the jump island address would be embedded as a data pointer (e.g., an info
 * table pointer in a closure), causing the GC to interpret jump island memory
 * as a valid info table — leading to "strange closure type" crashes.
*
 * We detect data references using ELF_ST_TYPE(sym.st_info) from the symbol
 * table. When a non-function symbol overflows 32-bit range, we emit a clear
 * error message instead of silently creating a corrupt jump island.
 *
 * Workarounds for data reference overflow:
 *   - Use shared Haskell libraries (dynamically-linked GHCi)
 *   - Compile with -fPIC -fexternal-dynamic-refs (generates GOT-based
 *     relocations that can handle arbitrary distances)
 *   - Use +RTS -xp (maps loaded objects into low memory)
 *
```

hsyl20 commented on rts/linker/Elf.c:
Nice! We should probably upstream this.

Is it an issue for tables-next-to-code where a symbol is both data and code (entry code)? Shouldn't we use two symbols instead of one to trigger this error when appropriate?

hsyl20 commented on rts/rts.cabal:
Sub-libraries are such a kludge. I'm not convinced they are worth all the added complexity

hsyl20 commented on testsuite/driver/testlib.py:
Do you remember why it failed? We need some motivation to upstream it. It looks reasonable but a concrete example would help.

hsyl20 commented on testsuite/driver/testlib.py:
Ah, I guess the code above is to handle this case. Maybe not good for upstreaming then.

hsyl20 commented on utils/ghc-iserv/ghc-iserv.cabal.in:
The Darwin/OSX `-flat_namespace` option should be passed in `GHC.Runtime.Interpreter.C` too.

(we already pass `-Wl,--export-dynamic` on ELF targets there)

hsyl20 commented on Makefile:
Do we still need to distribute iserv programs now that GHC can build them on demand?

angerman commented on rts/rts.cabal:
I'm open to other solutions that play nice with cabal. I'm not super happy about the sublibs, but they seem to be the least worst of the pack.

angerman commented on rts/linker/Elf.c:
I just remember debugging this and scratching my head wtf was going on. Hence this at least gives a bit clearer error message. Two symbols might be nice, although it would lead to even more symbol explosion. Windows also struggles with this data/code issue massively. Ultimately, I'm not sure TNTC is a feature worth keeping. (it also inflates object sizes quite a bit by putting a lot of redundant data into the objects).

luite commented on utils/ghc-iserv/ghc-iserv.cabal.in:
does this mean that on Windows the dynamic build doesn't work?

angerman commented on rts/linker/Elf.c:
This is good!

angerman commented on testsuite/driver/testlib.py:
Not sure. Making it more robust might be a good idea (and reduce the stable-haskell ghc/upstream drift).

angerman commented on utils/ghc-iserv/ghc-iserv.cabal.in:
I don't think it has ever? And yes, this is a stupid limitation. But the issue is more about how many symbols we actually try to have visible.

angerman commented on Makefile:
For now yes. In the future I hope not.
